test/integration/http_integration.cc

```diff
@@ -1274,7 +1279,104 @@ void HttpIntegrationTest::testUpstreamDisconnectWithTwoRequests() {
   test_server_->waitForCounterGe("cluster.cluster_0.upstream_rq_200", 2);
 }
 
-void HttpIntegrationTest::testTwoRequests() {
+// This filter exists to synthetically test network backup by faking TCP
+// conneciton back-up when an encode is finished.
+// Allows regression tests for https://github.com/envoyproxy/envoy/issues/4541
+// TODO(alyssawilk) move this somewhere general and turn it up for more tests.
+class TestPauseFilter : public PassThroughFilter {
+public:
+  // A wrapper around a TimerPtr which can be passed to the timer on creation,
+  // so it can delete itself when called.
+  struct SelfOwningTimer {
+    Event::TimerPtr timer_;
+  };
+
+  // Pass in a global "is this the first stream to finish encoding" to ensure
+  // the Network::Connection is "blocked" exactly once.
+  TestPauseFilter(absl::Mutex& encode_lock, bool& first_encode_ref)
+      : encode_lock_(encode_lock), first_encode_ref_(first_encode_ref) {}
+
+  Http::FilterDataStatus encodeData(Buffer::Instance& buf, bool end_stream) override {
+    absl::WriterMutexLock m(&encode_lock_);
+    // If this is the first time any stream has finished encoding, force the
+    // connection to a high watermark state, and set an alarm to force low
+    // watermark state.
+    if (end_stream && first_encode_ref_) {
+      forceHighWatermark();
+    }
+    return PassThroughFilter::encodeData(buf, end_stream);
+  }
+
+  void forceHighWatermark() {
+    first_encode_ref_ = false; // Make sure we only do this once.
+    Network::ConnectionImpl* connection_impl = connection();
+    connection_impl->onHighWatermark(); // Force high watermark state.
+
+    // Unfortunately this filter is about to self destruct, so the timer
+    // to lower the watermark state must outlive the filter.
+    //
+    // Create a "self-owning" timer and pass it an absolute pointer to the
+    // connection. The timer will delete itself on destruction.
+    //
+    // It is up to the users of this filter to make sure the connection_impl
+    // will outlive the timer.
+    SelfOwningTimer* timer_container = new SelfOwningTimer;
```

Comment:
nit: in C++14 you can actually make this a unique_ptr and capture it into the lambda below using "generalized lambda capture". Then you can avoid the raw new/delete which is kind of nice.

Reply:
TIL :-)